File: nexIRC.Client/Classes/Communications/StatusSocket.vb
Option Explicit On
Option Strict Off
Imports nexIRC.clsIrcNumerics
Imports nexIRC.Modules
Imports nexIRC.Sockets

Namespace Classes.Communications
    Public Enum SocketType
        Status = 1
        Ident = 2
    End Enum

    Public Class StatusSocketObject
        Public Event DataArrival(data As String)
        Private WithEvents _socket As AsyncSocketController
        Private _statusID As Integer
        Private _invoke As Form
        Private _socketType As SocketType = SocketType.Status

        Public WriteOnly Property SetSocketType() As SocketType
            Set(value As SocketType)
                _socketType = value
            End Set
        End Property

        Public WriteOnly Property PassSocket(invokeForm As Form) As AsyncSocketController
            Set(value As AsyncSocketController)
                _socket = value
                _invoke = invokeForm
            End Set
        End Property

        Public Function Connected() As Boolean
            If (_socket IsNot Nothing) Then
                Return _socket.Connected
            Else
                Return False
            End If
        End Function

        Public Sub NewSocket(ByVal id As Integer, ByVal form As Form)
            _socket = New AsyncSocketController(form)
            _statusID = id
            _invoke = New Form
            _invoke = form
        End Sub

        Public Sub SendSocket(ByVal data As String)
            If (Connected() = True) Then _socket.Send(data & Environment.NewLine)
        End Sub

        Public Sub CloseSocket()
            If Connected() = True Then _socket.Close()
        End Sub

        Public Sub ConnectSocket(ByVal ip As String, ByVal port As Long)
            If Connected() = False Then _socket.Connect(ip, port)
        End Sub

        Private Sub socket_CouldNotConnect(SocketID As String) Handles _socket.CouldNotConnect
            lStrings.ProcessReplaceString(_statusID, eStringTypes.sCOULD_NOT_CONNECT, lStatus.ServerDescription(_statusID))
        End Sub

        Private Sub lSocket_socketConnected(ByVal socketID As String) Handles _socket.SocketConnected
            lStatus.ConnectEvent(_statusID)
        End Sub

        Public Function Port() As Integer
            Return _socket.Port
        End Function

        Private Sub lSocket_socketDataArrival(ByVal socketID As String, ByVal socketData As String, ByVal bytes() As Byte, ByVal lByteRead As Integer) Handles _socket.SocketDataArrival
            Select Case _socketType
                Case SocketType.Status
                    lProcessNumeric.lIrcNumericHelper.ProcessDataArrival(_statusID, socketData)
                Case SocketType.Ident
                    RaiseEvent DataArrival(socketData)
            End Select
        End Sub

        Private Sub socket_socketDisconnected(ByVal socketId As String) Handles _socket.SocketDisconnected
            lStatus.CloseStatusConnection(_statusID, False)
        End Sub
    End Class
End Namespace

File: nexIRC.Socket/StatusSocket.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Windows.Forms;
//nexIRC 3.0.31
//Sunday, Oct 4th, 2014 - guideX
// ERROR: Not supported in C#: OptionDeclaration
//using nexIRC.clsIrcNumerics;
//using nexIRC.Modules;
//using nexIRC.Sockets;

namespace nexIRC.Sockets {
    public class StatusSocket {
        public event DataArrivalEventHandler DataArrival;
        public delegate void DataArrivalEventHandler(string data);
        private AsyncSocket withEventsField_socket;
        private AsyncSocket socket {
            get { return withEventsField_socket; }
            set {
                if (withEventsField_socket != null) {
                    withEventsField_socket.CouldNotConnect -= socket_CouldNotConnect;
                    withEventsField_socket.SocketConnected -= lSocket_socketConnected;
                    withEventsField_socket.SocketDataArrival -= lSocket_socketDataArrival;
                    withEventsField_socket.SocketDisconnected -= socket_socketDisconnected;
                }
                withEventsField_socket = value;
                if (withEventsField_socket != null) {
                    withEventsField_socket.CouldNotConnect += socket_CouldNotConnect;
                    withEventsField_socket.SocketConnected += lSocket_socketConnected;
                    withEventsField_socket.SocketDataArrival += lSocket_socketDataArrival;
                    withEventsField_socket.SocketDisconnected += socket_socketDisconnected;
                }
            }
        }
        private int statusId;
        private delegate void StringDelegate(string data);
        private delegate void IntegerDelegate(int @int);
        private delegate void DataArrivalDelegate(int id, string data);
        private delegate void DisconnectDelegate(int id, bool closeSocket);
        private Form _invoke;

        private SocketType _socketType = SocketType.Status;
        public enum SocketType {
            Status = 1,
            Ident = 2
        }

        public SocketType SetSocketType {
            set {
                try {
                    _socketType = value;
                } catch (Exception ex) {
                    throw ex;
                }
            }
        }

        public AsyncSocket PassSocket {
            set {
                try {
                    _invoke = invokeForm;
                    socket = value;
                } catch (Exception ex) {
                    throw ex;
                }
            }
        }

        public bool Connected() {
            try {
                return socket.Connected;
            } catch (Exception ex) {
                throw ex;
            }
        }

        public string ReturnLocalIp() {
            string functionReturnValue = null;
            try {
                functionReturnValue = socket.ReturnLocalIp();
            } catch (Exception ex) {
                throw ex;
                return null;
            }
            return functionReturnValue;
        }

        public long ReturnLocalPort() {
            try {
                if ((Connected() == true)) {
                    return socket.ReturnLocalPort();
                }
                return 0;
            } catch (Exception ex) {
                throw ex;
            }
        }

        public void NewSocket(int id, Form form) {
            try {
                socket = new AsyncSocket();
                statusId = id;
                _invoke = new Form();
                _invoke = form;
            } catch (Exception ex) {
                throw ex;
            }
        }

        public void SendSocket(string data) {
            try {
                if ((Connected() == true))
                    socket.Send(data + Environment.NewLine);
            } catch (Exception ex) {
                throw ex;
            }
        }

        public void CloseSocket() {
            try {
                if (Connected() == true)
                    socket.Close();
            } catch (Exception ex) {
                throw ex;
            }
        }

        public void ConnectSocket(string ip, long port) {
            try {
                if (Connected() == false)
                    socket.Connect(ip, port);
            } catch (Exception ex) {
                throw ex;
            }
        }

        public void CouldNotConnect(string data) {
            try {
                lStrings.ProcessReplaceString(statusId, eStringTypes.sCOULD_NOT_CONNECT, lStatus.ServerDescription(statusId));
            } catch (Exception ex) {
                throw ex;
            }
        }

        private void socket_CouldNotConnect(string SocketID) {
            try {
                StringDelegate couldNotConnectEvent = new StringDelegate(CouldNotConnect);
                _invoke.Invoke(couldNotConnectEvent, SocketID);
            } catch (Exception ex) {
                throw ex;
            }
        }

        private void lSocket_socketConnected(string socketID) {
            try {
                IntegerDelegate connectEvent = new IntegerDelegate(lStatus.ConnectEvent);
                _invoke.Invoke(connectEvent, statusId);
            } catch (Exception ex) {
                throw ex;
            }
        }

        public void DataArrivalProc(int id, string data) {
            try {
                if (DataArrival != null) {
                    DataArrival(data);
                }
            } catch (Exception ex) {
                throw ex;
            }
        }

        private void lSocket_socketDataArrival(string socketID, string socketData, byte[] bytes, int lByteRead) {
            try {
                switch (_socketType) {
                    case SocketType.Status:
                        DataArrivalDelegate processDataArrival = new DataArrivalDelegate(lProcessNumeric.lIrcNumericHelper.ProcessDataArrival);
                        _invoke.Invoke(processDataArrival, statusId, socketData);
                        break;
                    case SocketType.Ident:
                        DataArrivalDelegate processDataArrival = new DataArrivalDelegate(DataArrivalProc);
                        _invoke.Invoke(processDataArrival, statusId, socketData);
                        break;
                }
            } catch (Exception ex) {
                throw ex;
            }
        }

        private void socket_socketDisconnected(string socketId) {
            try {
                DisconnectDelegate disconnectEvent = new DisconnectDelegate(lStatus.CloseStatusConnection);
                if (Connected() == true)
                    _invoke.Invoke(disconnectEvent, statusId, false);
            } catch (Exception ex) {
                //Throw ex
            }
        }
    }
}
